hepcrawl/middlewares.py

```diff
@@ -34,3 +45,113 @@ def process_exception(self, request, exception, spider):
+
+    * ``crawl_once_value`` - a value to store in DB. By default, timestamp
+      is stored for Http/Https requests and last-modified is stored for FTP/File requests.
+    * ``crawl_once_key`` - unique file name is used.
+
+    Settings:
+
+    * ``CRAWL_ONCE_ENABLED`` - set it to False to disable middleware.
+      Default is True.
+    * ``CRAWL_ONCE_PATH`` - a path to a folder with crawled requests database.
+      By default ``.scrapy/crawl_once/`` path is used; this folder contains
+      ``<spider_name>.sqlite`` files with databases of seen requests.
+    * ``CRAWL_ONCE_DEFAULT`` - default value for ``crawl_once`` meta key
+      (False by default). When True, all requests are handled by
+      this middleware unless disabled explicitly using
+      ``request.meta['crawl_once'] = False``.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(HepcrawlCrawlOnceMiddleware, self).__init__(*args, **kwargs)
+
+    # def process_spider_output(self, response, result, spider):
+    #     for r in result:
+    #         yield r
+    #
+    #     if response.meta.get('crawl_once', self.default):
+    #         key = self._get_key(response.request)
+    #         self.db[key] = response.meta.get('crawl_once_value')
+    #         self.stats.inc_value('crawl_once/stored')
+
+    @staticmethod
+    def _is_newer(this_timestamp, than_this_timestamp, scheme):
+        if scheme in ['ftp', 'file']:
+            return this_timestamp > than_this_timestamp
+
+    def _has_to_be_crawled(self, request, spider):
+        request_db_key = self._get_key(request)
+
+        if request_db_key not in self.db:
+            return True
+
+        new_request_timestamp = self._get_timestamp(request, spider)
+        parsed_url = urlparse(request.url)
+        if self._is_newer(
+            new_request_timestamp,
+            self.db.get(key=request_db_key),
+            scheme=parsed_url.scheme,
+        ):
+            return True
+
+        return False
+
+    def process_request(self, request, spider):
+        request.meta['crawl_once_key'] = os.path.basename(request.url)
+        request.meta['crawl_once_value'] = self._get_timestamp(request, spider)
+
+        if not request.meta.get('crawl_once', self.default):
+            return
+
+        if not self._has_to_be_crawled(request, spider):
+            self.stats.inc_value('crawl_once/ignored')
+            raise IgnoreRequest()
+
+    @staticmethod
+    def _get_timestamp(request, spider):
+        parsed_url = urlparse(request.url)
+        full_url = request.url
+        if parsed_url.scheme == 'ftp':
+            ftp_host, params = ftp_connection_info(spider.ftp_host, spider.ftp_netrc)
+            ftp = FTP(
+                host=ftp_host,
+                user=params['ftp_user'],
+                passwd=params['ftp_password'],
+            )
+            file_path = full_url.replace(
+                '{0}://{1}/'.format(parsed_url.scheme, ftp_host),
+                '',
+            )
```

Comment:
You can move this to a method called `get_relative_path` or similar so it's clearer.